test/mem.test.js

```diff
@@ -288,6 +287,330 @@ describe('mem-store tests', function () {
       })
     })
   })
+
+  describe('internal utilities', () => {
+    const mem_store = seneca.export('mem-store')
+    const { intern } = mem_store.init
+
+    describe('find_one_doc', () => {
+      const ent_base = 'sys'
+      const ent_name = 'product'
+
+
+      describe('no such entities exist', () => {
+        let entmap
+
+        beforeEach(async () => {
```

Comment:
seems a little OTT :)
maybe let entmap = {} in the test itself is sufficient?
the rule of thumb for DRY is >2 instances

Reply:
in this case `let entmap = {}` in the test itself does exactly that - does it not?

it's ok to repeat a little setup code in tests

Reply:
also you're using beforeEach for individual tests - this would make more sense if it was for multiple tests - only go DRY once there is a need
The fewer lines of code, in general, the better

Reply:
> also you're using beforeEach for individual tests

I believe, using `beforeEach` allows a developer to save time and prevent bugs, when more tests are added. It's easy to forget to update a `before` hook to `beforeEach`, and it also takes extra time to do that - whereas quite often the performance overhead of `beforeEach` over `before` is marginal. As such, `beforeEach` is more "code-scalable" and provides better DX.

Reply:
Please don't do this. We need to apply the YAGNI heuristic here. If multiple tests cases are needed, and they need preparation (that is more than just creating empty objects), then it makes sense. The beforeEach statements severely occlude the meaning of the test. Each line of code must pay it's way.

Reply:
Will do.

However, I would argue that the following may be subjective:
> The `beforeEach` statements severely occlude the meaning of the test.

To me, personally, it's the opposite: `before` implies that no matter how many tests are added - the hook can only ever run once for all tests. The `beforeEach` hook, on the other hand, suggests to me that the code should run for each test.

Reply:
Another problem with mixing `before` and `beforeEach` is that `before` hooks always run before `beforeEach` hooks. Which means the following will not work:
```javascript
beforeEach(() => {
  console.log('this hook actually runs second')
  seneca = makeSenecaForTest()
})

describe('foo', () => {
  before(() => {
    console.log('this hook actually runs first')
    return seneca.make('user').save$()
  })

  // ...
})
```

Reply:
in that case, make `seneca` a local to the test:

`let seneca = makeSenecaForTest()`

now there's no way it can ever be dirty, and the repeated code is put inside the util function